examples/mdetr/contrastive_alignment_loss.py

```diff
@@ -0,0 +1,113 @@
+                indices (List[Tuple[Tensor, Tensor]]): A list of size batch_size,
+                containing tuples of (index_i, index_j) where:
+                    - index_i is the indices of the selected predictions (in order)
+                    - index_j is the indices of the corresponding selected targets
+                For each batch element, it holds:
+                    len(index_i) = len(index_j) = min(num_queries, num_target_boxes)
+            num_boxes (int): Normalization factor. Should equal the average number of
+                boxes per local batch.
+            tokenized (BatchEncoding): Tokenized output from a transformers fast tokenizer.
+                Used for token lookup based on character positions.
+            temperature (float): Scaling factor used in calculating the logits.
+                Default: 0.07
+    """
+
+    logits = (
+        torch.matmul(projected_queries, projected_tokens.transpose(-1, -2))
+        / temperature
+    )  # BS x (num_queries) x (num_tokens)
+
+    positive_map = construct_positive_map(logits, target_tokens, indices, tokenized)
+
+    positive_logits = -logits.masked_fill(~positive_map, 0)
+    negative_logits = logits
+
+    # Calculate the contrastive loss for all objects
+    boxes_with_pos = positive_map.any(2)
+    pos_term = positive_logits.sum(2)
+    neg_term = negative_logits.logsumexp(2)
+    nb_pos = positive_map.sum(2) + 1e-6
+    box_to_token_loss = (
+        ((pos_term / nb_pos + neg_term)).masked_fill(~boxes_with_pos, 0).sum()
+    )
+
+    # Calculate the contrastive loss for all tokens
+    tokens_with_pos = positive_map.any(1)
+    pos_term = positive_logits.sum(1)
+    neg_term = negative_logits.logsumexp(1)
+    nb_pos = positive_map.sum(1) + 1e-6
+    tokens_to_boxes_loss = (
+        ((pos_term / nb_pos + neg_term)).masked_fill(~tokens_with_pos, 0).sum()
+    )
+
+    tot_loss = (box_to_token_loss + tokens_to_boxes_loss) / 2
+
+    return tot_loss / num_boxes
+
+
+def construct_positive_map(logits: Tensor, target_tokens, indices, tokenized):
+    # construct a map such that positive_map[k, i,j] = True iff query i is associated to token j in batch item k
+    # For efficency, the construction happens on CPU, then the whole matrix is transferred to GPU in one go.
+    positive_map = torch.zeros(logits.shape, dtype=torch.bool)
+    for i, ((idx_src, idx_tgt), tgt) in enumerate(zip(indices, target_tokens)):
+        cur_tokens = [tgt[j] for j in idx_tgt]
+        for j, tok_list in enumerate(cur_tokens):
+            for (beg, end) in tok_list:
+                beg_pos = tokenized.char_to_token(i, beg)
+                end_pos = tokenized.char_to_token(i, end - 1)
+                if beg_pos is None:
+                    try:
+                        beg_pos = tokenized.char_to_token(beg + 1)
+                        if beg_pos is None:
+                            beg_pos = tokenized.char_to_token(beg + 2)
+                    except Exception:
```

Comment:
why are we checking offsets of beg and end? is it to find multiple tokens that are associated with the query? then why only upto offset of 3?

Reply:
I think this is just some heuristic they use for error handling. Actually I just tested on the phrase grounding eval code and it never even hits this portion of the code (the first char_to_token call always works). So I will take it out to keep things simpler, if we find that we need it for VQA or something we can always add it back later on